File: crates/suspect-codegen/src/python_http/credential_env.py
"""Creation-time environment defaults for an explicitly configured source policy."""
from __future__ import annotations

from collections.abc import Mapping
from typing import final
from ._types import Credential


@final
class _OmittedAuth:
    __slots__ = ()

    def __repr__(self) -> str:
        return '<auth omitted>'


_OMITTED_AUTH = _OmittedAuth()


def _credentials(auth: Mapping[str, Credential] | None | _OmittedAuth) -> Mapping[str, Credential] | None:
    if not isinstance(auth, _OmittedAuth):
        return auth
    # Explicit arguments return above without accessing the process environment.
    # Importing the generated package defines this function and never calls it.
    try:
        import os
    except ImportError:
        return {}
    values: dict[str, str | None] = {}
    credentials: dict[str, Credential] = {}
    for name, variable in _BINDINGS:
        if variable not in values:
            try:
                value = os.getenv(variable)
            except Exception:
                # Environment access may be unavailable. Retain no exception or
                # value; ordinary operation security handles missing credentials.
                value = None
            values[variable] = value if type(value) is str and value else None
        supplied = values[variable]
        if supplied is not None:
            credentials[name] = supplied
    return credentials

File: crates/suspect-codegen/src/python_http/credential_env.rs
//! Optional constructor surface and runtime for the shared bound environment plan.
use crate::credential_env::CredentialEnvPlan;

pub(super) fn runtime(plan: &CredentialEnvPlan) -> String {
    let mut source = include_str!("credential_env.py").to_owned();
    source.push_str("\n_BINDINGS: tuple[tuple[str, str], ...] = (\n");
    for binding in plan.bindings() {
        source.push_str(&format!(
            "    ({}, {}),\n",
            super::native_examples::quote(binding.name()),
            super::native_examples::quote(binding.variable()),
        ));
    }
    source.push_str(")\n");
    source
}

// Private module aliases isolate policy helpers from source-allocated model,
// operation and group names; no public factory name is added to the allocator.
pub(super) const IMPORTS: &str = "from collections.abc import Mapping as _Mapping\nimport httpx as _httpx\nfrom ._types import Credential as _Credential\nfrom . import _credential_env as _credential_env\n";

pub(super) fn constructor(asynchronous: bool) -> String {
    let transport = if asynchronous {
        "AsyncBaseTransport"
    } else {
        "BaseTransport"
    };
    format!(
        "\n    def __init__(self, *, auth: _Mapping[str, _Credential] | None | _credential_env._OmittedAuth = _credential_env._OMITTED_AUTH, server_url: str | None = None,\n                 server: int | str = 0, server_variables: _Mapping[str, str] | None = None,\n                 document_url: str | None = None, auth_alternative: int | None = None,\n                 transport: _httpx.{transport} | None = None, max_response_bytes: int | None = None,\n                 max_capture_bytes: int = 4096, timeout: float | None = 30) -> None:\n        super().__init__(auth=_credential_env._credentials(auth), server_url=server_url, server=server,\n                         server_variables=server_variables, document_url=document_url, auth_alternative=auth_alternative,\n                         transport=transport, max_response_bytes=max_response_bytes, max_capture_bytes=max_capture_bytes, timeout=timeout)\n"
    )
}
